src/ucp/wireup/select.c

```diff
@@ -1426,29 +1427,94 @@ ucp_wireup_is_md_map_count_valid(ucp_context_h context, ucp_md_map_t md_map)
            (ucs_popcount(md_map) < UCP_MAX_OP_MDS);
 }
 
+static double ucp_wireup_get_lane_progress_overhead(ucp_worker_h worker,
+                                                    ucp_rsc_index_t rsc_index)
+{
+    ucp_worker_iface_t *wiface = ucp_worker_iface(worker, rsc_index);
+    uct_perf_attr_t perf_attr;
+
+    perf_attr.field_mask = UCT_PERF_ATTR_FIELD_OPERATION |
+                           UCT_PERF_ATTR_FIELD_PROGRESS_OVERHEAD;
+
+    /* We use AM_SHORT as a dummy operation because we don't know which
+        protocol will be selected (anyway it has no effect on the value of
+        progress_overhead). */
+    perf_attr.operation = UCT_EP_OP_AM_SHORT;
+
+    if (uct_iface_estimate_perf(wiface->iface, &perf_attr) != UCS_OK) {
+        /* Skip the overhead check if no value was retrieved */
+        return 0;
+    }
+
+    return perf_attr.progress_overhead;
+}
+
+static unsigned
+ucp_wireup_add_fast_lanes(const ucp_wireup_select_params_t *select_params,
+                          const ucp_wireup_select_info_t *sinfo_array,
+                          unsigned num_sinfo, ucp_worker_h worker,
+                          ucp_lane_type_t lane_type,
+                          ucp_wireup_select_context_t *select_ctx)
+{
+    unsigned num_lanes     = 0;
+    double min_overhead    = DBL_MAX;
+    ucp_context_h context  = worker->context;
+    const double max_ratio = 1. / context->config.ext.multi_lane_max_ratio;
+    ucs_status_t status;
+    int show_error;
+    unsigned sinfo_index;
+    double overhead;
+
+    /* Iterate over all elements and calculate minimum progress overhead */
+    for (sinfo_index = 0; sinfo_index < num_sinfo; ++sinfo_index) {
+        overhead     = ucp_wireup_get_lane_progress_overhead(
+                worker, sinfo_array[sinfo_index].rsc_index);
+        min_overhead = ucs_min(overhead, min_overhead);
+    }
+
+    /* Compare each element to minimum progress overhead and filter only fast lanes */
+    for (sinfo_index = 0; sinfo_index < num_sinfo; ++sinfo_index) {
+        overhead = ucp_wireup_get_lane_progress_overhead(
+                worker, sinfo_array[sinfo_index].rsc_index);
+        if ((min_overhead > 0) && ((min_overhead / overhead) < max_ratio)) {
```

Comment:
Something seems wrong with the logic hear.
`min_overhead` will be `0` if there's at least 1 "fast" iface, so this `if` would never happen?
Also, I would care more about checking the denominator (`overhead`) here to prevent division by 0.

Reply:
correct, it was some leftovers from prev code.

Reply:
min_overhead is always bigger than overhead, so min_overhead > 0 also implies overhead > 0.
if min_overhead is 0 than we don't want to filter anything, so the condition will never be true.

Reply:
"min_overhead is always bigger than overhead" - are you sure?
Line 1472:
`min_overhead = ucs_min(overhead, min_overhead);`